Add more details and a concrete proposal for KEP-5

jbeda opened the pull request:
This is a concrete proposal on how we can implement a community site.  I'm looking forward to a healthy discussion around this proposal.

heckj commented:
I’ll try and join the contrib-ex call, and in the meantime I’ll cobble some quick thoughts here so they’re exposed:

- I’m the first to admit the k8s.io has flaws and can be damned awkward. That said, adding to the number of different sites when the problem statement says “there are too many places for information” strikes as trying to solve the problem by creating a new “one ring to rule them all”. _insert comic from XKCD about standards here_
- I totally get the workflow concern, and think it can be directly alleviated with the recent adoption of the Prow plugins that SIG Docs went through the pain of adopting. We can identify approvers for a sub-tree portion, and I don’t think it’s unreasonable to have a different set of folks managing a subset of documentation to expand the support capabilities of making excellent content available. My own opinion is that it also doesn’t need to abide by the same standards, depending on where you are on the site (would be nice, but not critical - and recognizing the manpower is the scarcest resource here)
  - a side comment to this is the ongoing-but-not-yet-merged work to add the Blog into the k8s.io site, which will also require a similiar workflow for the people posting to the blog content.
- I think there’s a potential misunderstanding of the branching structure of the docs - the majority of content on docs **is** an evergreen, hosted-to-master site - and does not “branch and lock” like typical software releases. That’s been a confusion point for other contributors - we branch as a means of keeping a copy-from-version, but don’t update those branches - and we branch as a “pre-load this content for the upcoming release” - with the effort made to not expose information before it’s ready, but the main content is always live and updated.
- We already have needs for tooling improvements in a variety of forms - the “hosting the blog in k8s.io” is one, the importing of various generated markdown another. To that end, I would argue that adding content tooling to support a single location and resolving the relevant indexing issues (indexing problems seemed to be implied in the KEP, but I wasn’t sure if it was really an issue) would be a far greater payout on investment than building another, alternate site and splitting the labor efforts.

Bradamant3 commented on keps/sig-contributor-experience/0005-community-site.md:
```diff
@@ -97,14 +114,47 @@ Hopefully the content previously contained in [umbrella issues][] will be tracke
+This is not preferred for a variety of reasons.
+
+* **Workflow.** Currently there is quite a bit of process for getting things merged into the main site.
+  That process involves approval from someone on SIG-Docs from an editorial point of view along with approval for technical accuracy.
```
We're working on streamlining this process and getting PRs through more quickly. Many PRs currently get merged without two sets of review; technical accuracy is at issue only for, well, highly technical topics.

zacharysarah replied:
It's worth pointing out that an OWNERS file provides directory-level ownership (hypothetically, in kubernetes/website, docs/contrib/*) and can disallow permissions inheritance (`no_parent_owners: true`) to limit and isolate automation assignments. 

All of which is to say that it's possible to achieve review isolation on the main site.

Bradamant3 commented on keps/sig-contributor-experience/0005-community-site.md:
```diff
@@ -97,14 +114,47 @@ Hopefully the content previously contained in [umbrella issues][] will be tracke
+
+The main alternative here is to build something that is integrated into the user facing kubernetes.io site.
+This is not preferred for a variety of reasons.
+
```
I address specific points about the /website flow below without necessarily trying to make an argument in favor of integrating with the current docs site; I think it's helpful to clear up the issues regardless. (Also it's useful for sig-docs to see what misperceptions about docs flow and tooling persist.)

Bradamant3 commented on keps/sig-contributor-experience/0005-community-site.md:
```diff
@@ -97,14 +114,47 @@ Hopefully the content previously contained in [umbrella issues][] will be tracke
+  The two stage approval slows down contributions and creates a much larger barrier than the current markdown based flow.
+  In addition, SIG-Docs is already stretched thin dealing with the (more important) user facing content that is their main charter.
+* **Quality standards.** The bar for the user facing site is higher than that of the community site.
+  Speed and openness of communication dominates for the community facing site.
```
Docs maintainers are also currently in the practice of judging just how much we need to meet arbitrary quality standards. We try to err on the side of getting useful content into the docs. IOW, we're flexible on this. We don't unilaterally impose an arbitrary set of content standards.

zacharysarah replied:
It would be helpful to consistently talk about a contributor site, rather than using "contributors" and "community" interchangeably. It's clear what contributors are, and why a site matters for them; it's less clear what a community is, and site needs even less so.

jbeda replied:
I’m happy to center this around “contributors” if that helps makes progress.  The reason that they are somewhat conflated here is that I see this as an extension and evolution of the “community” repo.  If this repo were called “kubernetes/contributors” then I’m guessing we’d be more consistent.

Bradamant3 commented on keps/sig-contributor-experience/0005-community-site.md:
```diff
@@ -97,14 +114,47 @@ Hopefully the content previously contained in [umbrella issues][] will be tracke
+* **Quality standards.** The bar for the user facing site is higher than that of the community site.
+  Speed and openness of communication dominates for the community facing site.
+  Our bar here is the current pile of Markdown.
+* **Versioning and branching.** Much of the main user facing site is lined up on Kubernetes releases.
```
As heckj points out in his overall comment, this isn't the case. Ongoing PRs to /website are made to master, which publishes immediately. Only release-specific docs are merged to release-specific branches, which are then merged to master when the release ships. IOW, we do not follow the branch/publish strategy of k8s/k8s

jbeda commented:
Lots of good conversations in the SIG-contribX meeting. I'll point to the video here as it is available.

Bradamant3 commented on keps/sig-contributor-experience/0005-community-site.md:
```diff
@@ -60,30 +60,47 @@ Specifically, building a site that can take advantage of the KEP metadata will b
 
-* Discover and bring together information from multiple orgs/repos.
+* Actively migrate information from multiple orgs/repos.
+  This should be a place that people in the community choose to use to communicate vs. being forced.
```
I agree about not forcing, and creating a place where people choose to communicate. But couldn't we also encourage folks to migrate information from other repositories? There's a huge information architecture issue here, and without wanting to impose requirements on anyone, using the site as an opportunity to organize content better could be a huge win.

Bradamant3 commented:
My individual comments are probably not all that helpful in moving toward a useful solution, but I hope maybe they contribute to clearing up some tangential confusion.

Overall, a separate site probably makes the most sense at this point. I wonder about an additional goal: could this site also help to gather the markdown source into the single community repo? I'm guessing it'll be a while before everything that we want winds up on the site. See my comment about non-goals.

jbeda commented:
Video of discussion at the ContribX meeting is here: https://youtu.be/iSWtqbRjE1o?t=2m30s

@parispittman I'd love to get some direction from you and @castrojo on where you want to take this next.

parispittman commented:
Next steps from meeting today:

- [ ] @jbeda to modify the proposal to include contributor vs community language and other nits
- [ ] ContribEx to socialize the KEP for a final call for comments with at least 72 hours time box to the following mailing lists - k-dev, k-contribex, sig-leads
- [ ] Incorporate any feedback / edits that may come in
- [ ] @jbeda to start on prototype

...iterate from there

jbeda commented:
@parispittman I've updated this PR with the changes that we've talked about. I also added a graduation requirement.

My preference on how to move forward: (1) you approve/merge this if you are happy with it and (2) I create another PR that changes this from `provisional` to `implementable`. We discuss any lingering concerns there and let it bake. It is up to you how widely you feel you need to broadcast to solicit feedback.  Keep in mind that the more you ask for opinions the more opinions you'll get.  The final call will ultimately be up to you as the `approver` for this KEP.

Once we get that in we can start work on prototyping. (I may get started before that but won't submit any PRs until the KEP is `implementable`.

At any point we can decide this is a failed experiment or decide there are better approaches and change. We aren't locked in for life.

Thanks!

parispittman commented:
Let's go ahead to the next stage where we'll call for comments there, @jbeda 

/lgtm
/approve
